scipy/optimize/_lsq_dogbox.py

```diff
@@ -0,0 +1,353 @@
+
+    if scaling == 'jac':
+        if issparse(J):
+            scale = np.asarray(J.power(2).sum(axis=0)).ravel()**0.5
+        else:
+            scale = np.sum(J**2, axis=0)**0.5
+        scale[scale == 0] = 1
+    else:
+        scale = scaling
+    scale_inv = 1 / scale
+
+    Delta = norm(x0 * scale, ord=np.inf)
+    if Delta == 0:
+        Delta = 1.0
+
+    on_bound = np.zeros_like(x0, dtype=int)
+    on_bound[np.equal(x0, lb)] = -1
+    on_bound[np.equal(x0, ub)] = 1
+
+    x = x0.copy()
+    step = np.empty_like(x0)
+    cost = 0.5 * np.dot(f, f)
+
+    if max_nfev is None:
+        max_nfev = x0.size * 100
+
+    termination_status = None
+    iteration = 0
+    step_norm = None
+    actual_reduction = None
+
+    if verbose == 2:
+        print_header()
+
+    while nfev < max_nfev:
+        if scaling == 'jac':
+            if issparse(J):
+                new_scale = np.asarray(J.power(2).sum(axis=0)).ravel()**0.5
+            else:
+                new_scale = np.sum(J**2, axis=0)**0.5
+            scale = np.maximum(scale, new_scale)
+            scale_inv = 1 / scale
+
+        if isinstance(J, LinearOperator):
+            g = J.rmatvec(f)
+        else:
+            g = J.T.dot(f)
+
+        active_set = on_bound * g < 0
+        free_set = ~active_set
+
+        g_free = g[free_set]
+        if np.all(active_set):
+            g_norm = 0.0
+            termination_status = 1
+        else:
+            g_norm = norm(g_free, ord=np.inf)
+            if g_norm < gtol:
+                termination_status = 1
+
+        if verbose == 2:
+            print_iteration(iteration, nfev, cost, actual_reduction,
+                            step_norm, g_norm)
+
+        if termination_status is not None:
+            return OptimizeResult(
+                x=x, fun=f, jac=J, cost=cost, optimality=g_norm,
+                active_mask=on_bound, nfev=nfev, njev=njev,
+                status=termination_status)
+
+        x_free = x[free_set]
+        l_free = lb[free_set]
+        u_free = ub[free_set]
+        scale_inv_free = scale_inv[free_set]
+
+        # Compute (Gauss-)Newton and Cauchy steps
+        if tr_solver == 'exact':
+            J_free = J[:, free_set]
+            newton_step = lstsq(J_free, -f)[0]
+            Jg = J_free.dot(g_free)
+        elif tr_solver == 'lsmr':
+            Jop = aslinearoperator(J)
+            # Here we compute lsmr step in scaled variables and then
+            # transform back to normal variables, if lsmr would give exact lsq
+            # solution this would be equivalent to not doing any
+            # transformations, but from experience it's better this way.
+            lsmr_op = lsmr_linear_operator(Jop, scale_inv, active_set)
+            newton_step = -lsmr(lsmr_op, f, **tr_options)[0][free_set]
+            newton_step *= scale_inv_free
+            g[active_set] = 0
+            Jg = Jop.matvec(g)
+        cauchy_step = -np.dot(g_free, g_free) / np.dot(Jg, Jg) * g_free
```

Comment:
This computation is not very robust, because in theory `Jg` can be zero. I guess I need to form a coefficients of a quadratic function first and the compute the minimizer (as done in trf).